### PR TITLE
Introduce FSDPv2

### PR DESCRIPTION
**Pull request by alanwaketan**
Summary:
This patch introduce a PoC of FSDPv2. The full design doc is here: [go/fsdp_v2](http://goto.google.com/fsdp_v2). A real world use case can be found: https://github.com/pytorch-tpu/transformers/tree/llama2-spmd-fsdp.

Test Plan:
python test/spmd/test_fsdp_v2.py

### CODE REVIEW

File: torch_xla/experimental/spmd_fully_sharded_data_parallel.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+        if isinstance(output, TensorLike):
+          real_output = output
+        # TODO: enhance the logic as real_output can be something else. use _extract_tensors_from_list
+        elif isinstance(output, tuple):
+          real_output = output[0]
+        else:
+          raise RuntimeError(
+              f"The output type is not supported: {type(output)}. Please provide your own shard_output callable.")
+
+        spmd.mark_sharding(real_output, mesh, _prepare_spmd_partition_spec(real_output))
```
Comment (jonb377) [inline torch_xla/experimental/spmd_fully_sharded_data_parallel.py:80]: It looks like we can use SPMDFullyShardedDataParallel to express data-parallel multislice by specifying a mesh like `('dcn', 'fsdp')`, but we'd need to combine the two axes when sharding the activations' batch axis.

We could achieve this by specifying a `shard_output` function, but what do you think of allowing users to override which axes to shard activations along in the default `shard_output_impl`? e.g. a new constructor parameter for `activation_sharding='fsdp'`, and allow it to be overridden to `activation_sharding=('dcn', 'fsdp')`

File: test/spmd/test_fsdp_v2.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+  @classmethod
+  def setUpClass(cls):
+    xr.use_spmd()
+    super().setUpClass()
+
+  def test_fsdp_v2(self):
+    model = self.SimpleLinear().to(xm.xla_device())
+    mesh = self._get_mesh((self.n_devices, 1), None, ('fsdp', 'tensor'))
+    model.fc1 = fsdp.SpmdFullyShardedDataParallel(model.fc1, mesh)
```
Comment (jonb377) [inline test/spmd/test_fsdp_v2.py:25]: Do we need to shard the individual layers? Since the full model is wrapped on L27

**Review by jonb377 — COMMENTED** [review]
Awesome stuff Jiewen! 👏

File: test/spmd/test_fsdp_v2.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    model.fc2 = fsdp.SpmdFullyShardedDataParallel(model.fc2, mesh)
+    model = fsdp.SpmdFullyShardedDataParallel(model, mesh)
```
[the remark this answers is not in the document]
Comment (yeounoh) [reply test/spmd/test_fsdp_v2.py:27]: Is there a way to wrap recursively into the module? This is a question maybe more for the usage.. do you need to wrap each layer and then the module?

File: torch_xla/experimental/spmd_fully_sharded_data_parallel.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+from torch._prims_common import TensorLike, TensorSequenceType
+
+import numpy as np
+
+import torch_xla
+import torch_xla.distributed.spmd as spmd
+
+def _prepare_spmd_partition_spec(param):
+  partition_spec = [None] * len(param.shape)
+  # Skip scalar tensors.
```
Comment (yeounoh) [inline torch_xla/experimental/spmd_fully_sharded_data_parallel.py:14]: "Skip scalar tensor and call it replicated."

File: torch_xla/experimental/spmd_fully_sharded_data_parallel.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+  # Only shard the 0th dimension of the parameter according to the
+  # fsdp axis of the mesh.
+  # TODO: should we shard on the maximal dim for param? Then we need
+  # another helper for the output.
+  partition_spec[0] = "fsdp"
+  return tuple(partition_spec)
+
+
+class SpmdFullyShardedDataParallel(nn.Module):
+
```
Comment (yeounoh) [inline torch_xla/experimental/spmd_fully_sharded_data_parallel.py:27]: Let's add some docstring / description

File: test/spmd/test_fsdp_v2.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (yeounoh) [reply test/spmd/test_fsdp_v2.py:27]: Ok, first wrap the submodules and then wrapping the outer module will take care of the rest.

**Review by yeounoh — APPROVED** [review]
Added some minor comments, thanks!

File: test/spmd/test_fsdp_v2.py (hunk 2 above)
Comment (yeounoh) [reply test/spmd/test_fsdp_v2.py:25]: I would also go a step further to test 2 cases
- wrap one of the inners and then the outer, check if all are wrapped; let's say the two wrapping uses different shardings and make sure the original sharding is unchanged.
- wrap the outer, and try and except when wrapping the inner

File: torch_xla/experimental/spmd_fully_sharded_data_parallel.py (hunk 1 above)
Comment (alanwaketan) [reply torch_xla/experimental/spmd_fully_sharded_data_parallel.py:80]: Oops, I missed that.

File: test/spmd/test_fsdp_v2.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (alanwaketan) [reply test/spmd/test_fsdp_v2.py:27]: Right, the auto-wrap will come later.

**Comment by alanwaketan** [comment]
@jonb377 and @yeounoh thanks for the quick review. A lot of the things will become more clear once the design doc is ready. Will keep you posted.

File: torch_xla/experimental/spmd_fully_sharded_data_parallel.py [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+  def __init__(self, module: nn.Module, mesh: spmd.Mesh, shard_output:Optional[Callable] = None):
```
Comment (JackCaoG) [inline torch_xla/experimental/spmd_fully_sharded_data_parallel.py:28]: do we need to take mesh from caller? If it is FSDP, would it be possible to just assume we just shard at 0th dimension for all devices if mesh is not provided?

Comment (alanwaketan) [reply torch_xla/experimental/spmd_fully_sharded_data_parallel.py:28]: The thing is that the mesh will need to share with the dataloader. And a global mesh is a must for SPMD. I was thinking maybe we can have an API to set a global mesh. @jonb377 @yeounoh

**Review by JackCaoG — APPROVED** [review]

File: torch_xla/experimental/spmd_fully_sharded_data_parallel.py [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+
+  def forward(self, *args: Any, **kwargs: Any) -> torch.Tensor:
+    output = self.module(*args, **kwargs)
+    # Need to shard the output of the forward to instruct the compiler
+    # to enforce the FSDP algorithm.
+    self._shard_output(output, self._mesh)
+    return output
+
+  def __getattr__(self, name: str) -> Union[torch.Tensor, nn.Module]:
+    """Forward missing attributes to wrapped module."""
```
[the remark this answers is not in the document]
Comment (alanwaketan) [reply torch_xla/experimental/spmd_fully_sharded_data_parallel.py:99]: I just copy & paste from the existing wrapper. lol. Will need to revisit it.

File: torch_xla/experimental/spmd_fully_sharded_data_parallel.py (hunk 4 above)
Comment (alanwaketan) [reply torch_xla/experimental/spmd_fully_sharded_data_parallel.py:14]: Fixed.

File: test/spmd/test_fsdp_v2.py [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+import torch_xla
+import torch_xla.runtime as xr
+import torch_xla.core.xla_model as xm
+import torch_xla.distributed.spmd as xs
+
+import test_xla_sharding_base
+import torch_xla.experimental.spmd_fully_sharded_data_parallel as fsdp
+
+
+class BasicShardingTest(test_xla_sharding_base.XlaShardingTest):
```
[the remark this answers is not in the document]
Comment (alanwaketan) [reply test/spmd/test_fsdp_v2.py:15]: Fixed.

File: test/spmd/test_fsdp_v2.py (hunk 2 above)
Comment (alanwaketan) [reply test/spmd/test_fsdp_v2.py:25]: I hope this use case is well explained in the design doc.

File: torch_xla/experimental/spmd_fully_sharded_data_parallel.py (hunk 5 above)
Comment (alanwaketan) [reply torch_xla/experimental/spmd_fully_sharded_data_parallel.py:27]: Done.

File: torch_xla/experimental/spmd_fully_sharded_data_parallel.py (hunk 1 above)
Comment (alanwaketan) [reply torch_xla/experimental/spmd_fully_sharded_data_parallel.py]: Let's do this as a follow up.

**Comment by alanwaketan** [comment]
I have addressed some of the comments and polish the test case a bit. Feel free to re-review it. I will add more test cases once I'm back.

**Comment by alanwaketan** [comment]
I'm merging it in order to catch the 2.2 backport cutoff.
